firmware/main.c

```diff
@@ -3,32 +3,128 @@
 PortData port_a;
 PortData port_b;
 
+// Indicates whether the SPI Daemon is listening for USB traffic
 volatile bool booted = false;
+// LED Chan: TCC1/WO[0]
+#define PWR_LED_TCC_CHAN 1
+// CC channel 0 on TCC instance 1
+# define PWR_LED_CC_CHAN 0
+// The maximum counter value was chosen to get a 2s period heartbeat
+#define MAX_COUNTER 0x16E30
+// We have 16 slices of a sine wave
+#define NUM_POINT_SLICES 0x10
+// Number of loop iterations in a single slice
+#define COUNTS_IN_SLICE MAX_COUNTER/NUM_POINT_SLICES
+// Evenly spaced points along a sine wave, shifted up by 1, scaled by 0.5
+// multiplied by the max counter value
+const float sin_wave_points[] = {
+    46872,
+    64809.13784181649,
+    80015.50904777585,
+    90176.08144786903,
+    93744,
+    90176.08144786903,
+    80015.50904777585,
+    64809.13784181649,
+    46872.00000000001,
+    28934.862158183518,
+    13728.490952224147,
+    3567.9185521309782,
+    0,
+    3567.9185521309732,
+    13728.490952224138,
+    28934.86215818348
+};
+// The current counter value of our timer
+volatile u32 counter = 0;
+// The index of what point value we have just passed
+volatile u8 current_point_index = 0;
```

Comment:
Almost everything in the firmware is an interrupt at the same priority, so very little needs to be `volatile`.